### PR TITLE
Add option to save on each training node

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

There is currently a problem when using `load_best_model_at_end=True` on a training with multiple nodes: the model is only saved on the main process (so the machine rank 0) and machine with other ranks can't see the saved model (unless the system uses some kind of shared storage). This PR adds a flag to enable the save on each node for that situation, and avoids the hard failure when the model to reload is not found.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1357,10 +1357,18 @@ def train(
...
+            best_model_path = os.path.join(self.state.best_model_checkpoint, WEIGHTS_NAME)
+            if os.path.exists(best_model_path):
+                # We load the model state dict on the CPU to avoid an OOM error.
+                state_dict = torch.load(best_model_path, map_location="cpu")
+                # If the model is on the GPU, it still works!
+                self._load_state_dict_in_model(state_dict)
+            else:
+                logger.warn(
+                    f"Could not locate the best model at {best_model_path}, if you are running a distributed training "
+                    "on multiple nodes, you should activate `save_on_each_node=True`."
```
Comment (stas00) [inline src/transformers/trainer.py:1370]: With the dual CLI/non-CLI API I'm never sure when to use `--arg` and when `arg` in the docs/help - perhaps it's safer to err on the longer version and non-CLI users will figure it out?

```suggestion
                    "on multiple nodes, you should activate `--save_on_each_node=True`."
```

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -183,6 +183,9 @@ class TrainingArguments:
         save_total_limit (:obj:`int`, `optional`):
             If a value is passed, will limit the total amount of checkpoints. Deletes the older checkpoints in
             :obj:`output_dir`.
+        save_on_each_node (:obj:`bool`, `optional`, defaults to :obj:`False`):
+            In multinode distributed training, whether save models or checkpoints on each node, or only on the main
```
Comment (stas00) [inline src/transformers/training_args.py:187]:
```suggestion
            When doing multi-node distributed training, whether to save models and checkpoints on each node, or only on the main one
```

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -456,6 +459,12 @@ class TrainingArguments:
             )
         },
     )
+    save_on_each_node: bool = field(
+        default=False,
+        metadata={
+            "help": "When doing a multinode distributed training, whether to save once per node or just once on the main node."
```
Comment (stas00) [inline src/transformers/training_args.py:465]:
```suggestion
            "help": "When doing multi-node distributed training, whether to save models and checkpoints on each node, or only on the main one"
```

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -937,6 +946,19 @@ def should_log(self):
             else:
                 return self.process_index == 0
 
+    @property
+    def should_save(self):
+        """
+        Whether or not the current process should save or more generally write to disk.
```
Comment (stas00) [inline src/transformers/training_args.py:952]:
```suggestion
        Whether or not the current process should write to disk, e.g., to save models and checkpoints.
```

**Review by stas00 — APPROVED** [review]
This is a very important fix - thank you, @sgugger!

File: src/transformers/training_args.py (hunk 3 above)
Comment (stas00) [reply src/transformers/training_args.py:465]: hmm, should we note that it'll use the same filenames? and thus not to activate when one uses a shared fs?

Comment (sgugger) [reply src/transformers/training_args.py]: I'll add it in the docstring.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -1357,10 +1357,18 @@ def train(
...
+            if os.path.exists(best_model_path):
+                # We load the model state dict on the CPU to avoid an OOM error.
+                state_dict = torch.load(best_model_path, map_location="cpu")
+                # If the model is on the GPU, it still works!
+                self._load_state_dict_in_model(state_dict)
+            else:
+                logger.warn(
+                    f"Could not locate the best model at {best_model_path}, if you are running a distributed training "
+                    "on multiple nodes, you should activate `--save_on_each_node`."
+                )
```
Comment (LysandreJik) [inline src/transformers/trainer.py:1371]: nitpick but this should maybe be a `logger.error` as it's quite a high priority log

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks @sgugger

File: src/transformers/trainer.py (hunk 5 above)
Comment (stas00) [reply src/transformers/trainer.py:1371]: or assert?

**Pull request merged**